File: app/src/main/java/com/forcetower/uefs/db_service/dao/EventDao.java
package com.forcetower.uefs.db_service.dao;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Transaction;

import com.forcetower.uefs.db_service.entity.Event;

import java.util.List;

/**
 * Created by João Paulo on 14/06/2018.
 */
@Dao
public abstract class EventDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insert(Event... events);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insert(List<Event> events);

    @Query("SELECT * FROM Event WHERE approved = 1 AND course_pointer = :course ORDER BY created_at DESC")
    public abstract LiveData<List<Event>> getAllEvents(int course);

    @Query("SELECT * FROM Event WHERE approved = 1 ORDER BY created_at DESC")
    public abstract LiveData<List<Event>> getAllEvents();

    @Delete
    public abstract void delete(List<Event> events);

    @Query("DELETE FROM Event WHERE approved = 1")
    public abstract void deleteAllApproved();

    @Query("DELETE FROM Event WHERE approved = 0")
    public abstract void deleteAllUnapproved();

    @Transaction
    public void deleteAndInsert(List<Event> events) {
        deleteAllApproved();
        insert(events);
        updateInsertionTime(System.currentTimeMillis()/1000);
    }

    @Query("UPDATE Event SET inserted_at = :createdAt WHERE approved = 1")
    public abstract void updateInsertionTime(long createdAt);

    @Query("UPDATE Event SET inserted_at = :createdAt WHERE approved = 0")
    public abstract void updateInsertionTimeUnapproved(long createdAt);

    @Query("SELECT * FROM Event WHERE approved = 0")
    public abstract LiveData<List<Event>> getAllUnapprovedEvents();

    @Transaction
    public void deleteAndInsertUnapproved(List<Event> events) {
        deleteAllUnapproved();
        insert(events);
        updateInsertionTimeUnapproved(System.currentTimeMillis()/1000);
    }

    @Query("UPDATE Event SET approved = 1 WHERE uuid = :uuid")
    public abstract void markEventApproved(String uuid);

    @Query("SELECT * FROM Event WHERE uuid = :uuid")
    public abstract LiveData<Event> getEvent(String uuid);
}

File: app/src/main/java/com/forcetower/unes/core/storage/database/dao/EventDao.kt
/*
 * Copyright (c) 2018.
 * João Paulo Sena <joaopaulo761@gmail.com>
 *
 * This file is part of the UNES Open Source Project.
 *
 * UNES is licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.forcetower.unes.core.storage.database.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import androidx.room.OnConflictStrategy.REPLACE
import com.forcetower.unes.core.model.event.*
import com.forcetower.unes.core.storage.database.accessors.SessionWithData
import com.forcetower.unes.core.storage.network.ServerSession
import timber.log.Timber

@Dao
abstract class EventDao {
    @Insert(onConflict = REPLACE)
    abstract fun insert(session: Session): Long

    @Transaction
    @Query("SELECT * FROM Session WHERE day_id = :day ORDER BY start_time ASC")
    abstract fun getSessionsFromDay(day: Int): LiveData<List<SessionWithData>>

    @Transaction
    open fun insertServerSessions(value: List<ServerSession>) {
        Timber.d("Requested insert of $value")
        value.forEach {
            insertTags(it.tags)
            insertSpeakers(it.speakers)
            val session = it.toSession()
            deleteIfPresent(session.uuid)
            val id = insert(session)

            it.tags.forEach { t ->
                assocTag(SessionTag(session = id, tag = t.uid))
            }

            it.speakers.forEach { s ->
                assocSpeaker(SessionSpeaker(session = id, speaker = s.uid))
            }
        }
    }



    @Query("SELECT * FROM Session WHERE uuid = :uuid")
    protected abstract fun getSessionWithUUID(uuid: String): Session?

    @Insert(onConflict = REPLACE)
    protected abstract fun insertTags(tags: List<Tag>)

    @Insert(onConflict = REPLACE)
    protected abstract fun insertSpeakers(speakers: List<Speaker>)

    @Insert(onConflict = REPLACE)
    protected abstract fun assocTag(tagged: SessionTag)

    @Insert(onConflict = REPLACE)
    protected abstract fun assocSpeaker(speaker: SessionSpeaker)

    @Query("DELETE FROM Session WHERE uuid = :sessionUUID")
    protected abstract fun deleteIfPresent(sessionUUID: String)
}